src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,3106 @@
+		next_free_dirfd = idx;
+
+	if (idx == last_dirfd) {
+		for (i = idx-1; i >= 0; i--) {
+			if (dir_list[i].dir_obj) {
+				last_dirfd = i;
+				break;
+			}
+		}
+	}
+
+	num_dirfd--;
+
+	fd_dup_pre = dir_list[idx].fd_dup_pre;
+	fd_dup_next = dir_list[idx].fd_dup_next;
+
+	if (fd_dup_pre >= 0) {
+		dir_list[fd_dup_pre].fd_dup_next = fd_dup_next;
+		if (fd_dup_next >= 0)
+			dir_list[fd_dup_next].fd_dup_pre = fd_dup_pre;
+	}
+	if ((fd_dup_pre == -1) && (fd_dup_next == -1))
+		ready_to_release = 1;
+
+	pthread_mutex_unlock(&lock_dirfd);
+
+	if (ready_to_release) {
+		rc = dfs_release(dir_obj);
+		assert(rc == 0);
+	}
+}
+
+static inline int
+Get_Fd_Redirected(int fd)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++) {
+		if (fd_dup2_list[i].fd_src == fd)
+			return fd_dup2_list[i].fd_dest;
+	}
+
+	if (fd >= 3)	{
+		return fd;
+	} else if (fd == 0) {
+		if (fd_stdin > 0)
+			return fd_stdin;
+	} else if (fd == 1) {
+		if (fd_stdout > 0)
+			return fd_stdout;
+	} else if (fd == 2) {
+		if (fd_stderr > 0)
+			return fd_stderr;
+	}
+
+	return fd;
+}
+
+static void
+init_fd_dup2_list(void)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		fd_dup2_list[i].fd_src = -1;
+		fd_dup2_list[i].fd_dest = -1;
+	}
+}
+
+static void
+free_fd_in_dup2_list(int fd)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		if (fd_dup2_list[i].fd_src == fd) {
+			fd_dup2_list[i].fd_src = -1;
+			fd_dup2_list[i].fd_dest = -1;
+		}
+	}
+}
+
+static int
+find_free_fd_dup2_list(void)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		if (fd_dup2_list[i].fd_src == -1)
+			return i;
+	}
+	printf("ERROR: num_fd_dup2ed >= MAX_FD_DUP2ED\n");
+	errno = EMFILE;
+	return (-1);
+}
+/**
+ * static int
+ * is_a_duped_fd(int fd)
+ * {
+ *	int i;
+ *
+ *	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+ *		if (fd_dup2_list[i].fd_src == fd)	{
+ *			return 1;
+ *		}
+ *	}
+ *	return 0;
+ * }
+ */
+static int
+query_fd_forward_dest(int fd_src)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		if (fd_src == fd_dup2_list[i].fd_src)
+			return fd_dup2_list[i].fd_dest;
+	}
+	return -1;
+}
+
+static int
+allocate_a_fd_from_kernel(void)
+{
+	struct timespec	times_loc;
+	char		file_name[128];
+
+	clock_gettime(CLOCK_REALTIME, &times_loc);
+	snprintf(file_name, sizeof(file_name) - 1, "dummy_%ld_%ld", times_loc.tv_sec,
+		 times_loc.tv_nsec);
+	return memfd_create("dummy", 0);
+}
+
+static void
+close_all_duped_fd(void)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		if (fd_dup2_list[i].fd_src >= 0)	{
+			fd_dup2_list[i].fd_src = -1;
+			new_close_common(real_close_libc, fd_dup2_list[i].fd_dest);
+		}
+	}
+}
+
+/**
+ *static int
+ * Query_Fd_Forward_Src(int fd_dest, int *Closed)
+ * {
+ *	int i;
+ *
+ *	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+ *		if (fd_dest == fd_dup2_list[i].fd_dest)	{
+ *			return fd_dup2_list[i].fd_src;
+ *		}
+ *	}
+ *	return -1;
+ * }
+ */
+static int
+open_common(int (*real_open)(const char *pathname, int oflags, ...), const char *szCallerName,
```

Comment:
Is the szCaller name used here?

Reply:
It was used for debugging purpose. Not in current code.